src/scitacean/transfer/ssh.py

```diff
@@ -73,50 +82,99 @@ def download_file(self, *, remote: str, local: Path):
         self._connection.get(remote=str(remote), local=str(local))
 
 
-class _ESSUploadConnection:
+class SSHUploadConnection:
     def __init__(
         self, *, connection: Connection, dataset_id: PID, remote_base_path: str
     ):
         self._connection = connection
         self._dataset_id = dataset_id
         self._remote_base_path = remote_base_path
 
+    @property
+    def _sftp(self) -> SFTPClient:
+        return self._connection.sftp()
+
     @property
     def source_dir(self) -> str:
         return os.path.join(self._remote_base_path, self._dataset_id.pid)
 
     def remote_path(self, filename) -> str:
         return os.path.join(self.source_dir, filename)
 
-    def upload_files(self, *files: File) -> List[File]:
-        """Upload files to the remote staging folder."""
-        return [
-            file.uploaded(
-                remote_path=self.upload_file(
-                    remote=file.remote_path, local=file.local_path
-                )
-            )
-            for file in files
-        ]
+    def _make_source_folder(self):
+        try:
+            self._sftp.mkdir(self.source_dir)
+        except OSError as exc:
+            raise FileUploadError(
+                f"Failed to create source folder: {exc.args}"
+            ) from None
 
-    def upload_file(self, *, remote: Union[str, Path], local: Union[str, Path]) -> str:
-        """Upload a file to the remote staging folder."""
-        remote_path = self.remote_path(remote)
+    def upload_files(self, *files: File) -> List[File]:
+        """Upload files to the remote folder."""
+        self._make_source_folder()
+        uploaded = []
+        try:
+            for file in files:
+                uploaded.append(self._upload_file(file))
+        except Exception:
+            self.revert_upload(*uploaded)
+            raise
+        return uploaded
+
+    def _upload_file(self, file: File) -> File:
+        remote_path = self.remote_path(file.remote_path)
         get_logger().info(
-            "Uploading file %s to %s on host %s", local, remote, self._connection.host
+            "Uploading file %s to %s on host %s",
+            file.local_path,
+            remote_path,
+            self._connection.host,
+        )
+        self._sftp.put(remotepath=remote_path, localpath=str(file.local_path))
+        st = self._sftp.stat(remote_path)
+        self._validate_upload(file, st)
+        return file.uploaded(
+            remote_gid=st.st_gid,
+            remote_uid=st.st_uid,
+            remote_creation_time=st.st_mtime,
+            remote_perm=st.st_mode,
         )
 
-        self._connection.run(f"mkdir -p {self.source_dir}", hide=True)
+    def _validate_upload(self, file: File, st: SFTPAttributes):
+        if file.size != st.st_size:
```

Comment:
Good question. I don't know. As I understand it, this code just places the bytes in a new file. So the only way this can change is if the OS tags on some metadata which is included in the file size. I am not aware of any filesystem that does this. Are you?

Reply:
I'm going to try uploading from Windows. If that works, I'd leave the check until it causes problems.